include/dem/insertion.h

```diff
@@ -54,30 +55,72 @@ template <int dim>
 class Insertion
 {
 public:
-  Insertion()
+  Insertion<dim>()
   {}
 
   virtual ~Insertion()
   {}
 
   /**
-   * Carries out the insertion of particles by discretizing and looping over the
-   * insertion region, finding the initial position of particles in the
-   * insertion region, their id and other initial properties of the particles.
-   * This virtual class serves as a template for all insertion functions.
+   * Carries out the insertion of particles. This is the base class of
+   * uniform_insertion and non_uniform_insertion classes.
    *
    * @param particle_handler The particle handler of particles which are being
    * inserted
    * @param triangulation Triangulation to access the cells in which the
    * particles are inserted
-   * @param property_pool Property pool of particles
-   * @param dem_parameters DEM parameters declared in the .prm file)
+   * @param dem_parameters DEM parameters declared in the .prm file
+   * @param inserted_this_step Number of particles that is going to be inserted
+   * at each insetion step. This value can change in the last insertion step to
+   * reach the desired number of particles
+   * @param nx Number of insertion points in the x direction
+   * @param ny Number of insertion points in the y direction
+   * @param nz Number of insertion points in the z direction
+   * @param remained_particles Remained number of particles that should be
+   * inserted in the following steps
    */
   virtual void
-  insert(Particles::ParticleHandler<dim> &particle_handler,
-         const Triangulation<dim> &       triangulation,
-         Particles::PropertyPool &        property_pool,
-         const DEMSolverParameters<dim> & dem_parameters) = 0;
+  insert(Particles::ParticleHandler<dim> &                particle_handler,
+         const parallel::distributed::Triangulation<dim> &triangulation,
+         const DEMSolverParameters<dim> &                 dem_parameters,
+         unsigned int &                                   inserted_this_step,
+         const unsigned int &                             nx,
+         const unsigned int &                             ny,
+         const unsigned int &                             nz,
```

Comment:
nx ny and nz should be given to the constructors of the sub classes. What happens if you have a class that does not use this notion to insert? The number of particles inserted and the insertion mechanism should be controlled internally. The DEM class should not be the one providing how much should be inserted this step. The insertion class should keep track of this.